KOGITO-536: Data-index-service Keycloak integration: GraphiQL (UI)

nmirasch opened the pull request:
jira related .. https://issues.redhat.com/browse/KOGITO-536
- Adds oidc keycloak configuration to allow data-index work as an service (working with bearer token) and as a webapp to serve the graphiql  redirecting to keycloak login page for authentication.

jstastny-cz commented on data-index/data-index-service/src/main/java/org/kie/kogito/index/auth/MultiTenantResolver.java:
```diff
@@ -0,0 +1,34 @@
+
+    @Override
+    public String resolve(RoutingContext context) {
+        if (context.request().path().startsWith("/graphiql")) {
```
so this decides which tenant is accessing the app based on the context, right. graphiql is a UI, thus web-app-tenant should be used, otherwise it will resort to the default config, is that correct?

nmirasch replied:
@jstastny-cz Correct! that's the idea, define a different oidc configuration for the ui (graphiql) and the graphql service. Regarding the keycloak authentication workflows are different in both cases.

nmirasch commented on data-index/data-index-service/src/test/java/org/kie/kogito/index/service/KeycloakIntegrationIndexingServiceIT.java:
```diff
@@ -66,27 +70,38 @@ public void testNoTokenProvided() {
     }
 
+    @Test
+    public void graphiqlLoginOnKeycloackServedTest(){
```
ups! thanks!

nmirasch commented on data-index/data-index-service/src/test/java/org/kie/kogito/index/service/KeycloakIntegrationIndexingServiceIT.java:
```diff
@@ -32,18 +33,25 @@
+
+    @BeforeAll
+    public static void setup() {
+        System.setProperty("quarkus.http.auth.policy.role-policy1.roles-allowed", "confidential");
```
@jstastny-cz yeah, I agree,  I have removed this. The purpose of this properties setting were regarding isolate  testing at my local, no need to be there.

nmirasch commented on data-index/data-index-service/src/test/java/org/kie/kogito/index/service/KeycloakIntegrationIndexingServiceIT.java:
```diff
@@ -66,27 +70,38 @@ public void testNoTokenProvided() {
+        // Returning keycloak login page
+        given().when().get("/graphiql/")
+                .then()
+                .statusCode(200);
```
@jstastny-cz Yeah, that value is the one returned by the quarkus oidc extension when it's configured as web-app. It returns request code for 'OK' and redirect to keycloak login.

cristianonicolai replied:
@nmirasch can we perhaps validate something in the body of the response that it actually redirected to Keycloack login page?

cristianonicolai replied:
@nmirasch I think just this one needs some improvement. A 200 could also be returned if the auth is disabled right? Anything that checks that it actually reached the login page.

cristianonicolai replied:
@nmirasch ^

nmirasch commented on data-index/data-index-service/src/main/resources/application.properties:
```diff
@@ -74,6 +74,13 @@ quarkus.oidc.auth-server-url=none
+%keycloak.quarkus.oidc.web-app-tenant.application-type=web-app
+
+# HTTP Security Configuration
+%keycloak.quarkus.http.auth.permission.authenticated.paths=/*
```
@jstastny-cz yeah! same comment as before here. That authentication configuration is passed as system prop at the pom.xml test declaration.

nmirasch commented:
@jstastny-cz updated, could you take a look, please? Thanks!!

Sgitario commented on data-index/data-index-service/src/test/java/org/kie/kogito/index/service/KeycloakIntegrationIndexingServiceIT.java:
```diff
@@ -32,18 +32,18 @@
                 .then()
-                .statusCode(404);
-
+                .statusCode(403);
```
I'm confused about the tests that cover the "/graphql" path. 
I ran the tests using "mvn clean verify" and I'm aware of these tests are working fine, but as far I can see in the VertxRouterSetup, we are not doing anything new about this path, right? 
Also, this path does not allow the GET method, but POST method... so no idea why is working. What am I missing?

jstastny-cz replied:
IMO the important fact is here that the user is not authorized to access the web context, details like methods allowed would be probably checked later on.

jstastny-cz commented on data-index/data-index-service/src/test/java/org/kie/kogito/index/service/KeycloakIntegrationIndexingServiceIT.java:
```diff
@@ -66,27 +62,38 @@ public void testNoTokenProvided() {
     }
 
+    @Test
+    public void graphiqlLoginOnKeycloakServedTest(){
```
See the comments below, that's the keycloak login page which returns 200.

jstastny-cz commented on data-index/data-index-service/src/test/java/org/kie/kogito/index/service/KeycloakIntegrationIndexingServiceIT.java:
```diff
@@ -32,18 +32,18 @@
 @QuarkusTestResource(InfinispanServerTestResource.class)
```
Not sure what you mean here. Do you mean like in `data-index/data-index-service/src/test/java/org/kie/kogito/index/InfinispanServerTestResource.java` ? Cause it's already in place.

Sgitario replied:
I mean to add a default image in case of the property is not set.

jstastny-cz replied:
I would argue against. Having configuration of the images both in pom and in code is not manageable easily, it could hide the fact that you've forgotten to set the property, the "somewhere defined default" could become outdated, etc. Anyway that should not be taken care of as part of this PR.

Sgitario replied:
Ok with taking care of this as part of another ticket. 
My against of having properties in the POM file is that is not IDE compatible (we can't run suites or single tests directly from the IDE) and for me, this is really useful for debugging.

jstastny-cz replied:
Well you can, you just need to set the property to your test runner. It's maybe just a question of taste, who likes what. Feel free to open a jira if you like.

Sgitario replied:
The problem is that a different test execution is used if we run the whole test suite, or one test or another (at least in Eclipse, not sure about Intellij or VS Code). So, I would need to add the properties in every test execution. Maybe, there could have another approaches.

Having these properties in a test.properties (or similar) and injecting this file into Quarkus would work when running the tests from Maven or any IDE. 

However, this is out of scope here. I wrote it here as I saw we're using default values for some other properties.

nmirasch replied:
@Sgitario I have to say that I have that default value for that props in my local environment to execute test from Intellij.. :-)
But maybe we can add there req in another jira just to see it it's convenient .

jstastny-cz commented on data-index/data-index-service/src/main/java/org/kie/kogito/index/vertx/VertxRouterSetup.java:
```diff
@@ -43,14 +48,37 @@
+        router.route().handler(FaviconHandler.create());
+        router.route("/").handler(ctx -> ctx.response().putHeader("location", "/graphiql/").setStatusCode(302).end());
         router.route("/graphql").handler(ApolloWSHandler.create(graphQL));
         router.route("/graphql").handler(GraphQLHandler.create(graphQL, new GraphQLHandlerOptions()));
```
That's a different endpoint. graphiql is web app, graphql is the service endpoint.

Sgitario replied:
The web application is also exposed via "graphql-ui" endpoint (this can be configured via the property "quarkus.vertx-graphql.ui.path"). But this is only working when running the app in dev mode (because the property "quarkus.vertx-graphql.ui.always-include" defaults to "false"). 

Therefore, as long as we only have one web app at runtime, we're good. Thanks!

nmirasch commented on data-index/data-index-service/src/main/java/org/kie/kogito/index/auth/MultiTenantResolver.java:
```diff
@@ -0,0 +1,34 @@
+import io.vertx.ext.web.RoutingContext;
+
+@ApplicationScoped
+public class MultiTenantResolver implements TenantResolver {
```
Hi @Sgitario, I thinks it's a good idea of making the graphiql path and oidc-tenant  given as configuration properties.Let me give it a try Thanks!

nmirasch commented on data-index/data-index-service/src/main/java/org/kie/kogito/index/vertx/VertxRouterSetup.java:
```diff
@@ -43,14 +48,37 @@
+            return ((QuarkusHttpUser) routingContext.user()).getSecurityIdentity().getCredential(AccessTokenCredential.class).getToken();
+        }
+        return "";
+    }
```
Thanks!

nmirasch commented:
@Sgitario I updated this code to include the kogito.dataindex.vertx-graphql.ui.path and kogito.dataindex.vertx-graphql.ui.tenant  properties to be able to configure that relationship by properties.
Could you take a look, please? Thanks!

Sgitario commented:
jenkins please retest this

Sgitario commented on data-index/data-index-service/src/main/java/org/kie/kogito/index/vertx/VertxRouterSetup.java:
```diff
@@ -39,18 +47,44 @@
-            router.route().handler(FaviconHandler.create());
-            router.route("/").handler(ctx -> ctx.response().putHeader("location", "/graphiql/").setStatusCode(302).end());
+        GraphiQLHandler graphiQLHandler = GraphiQLHandler.create(new GraphiQLHandlerOptions().setEnabled(true));
+        if (Boolean.TRUE.equals(authEnabled)) {
```
Should we cover these changes in VertxRouterSetup in an unit test?

nmirasch replied:
added

nmirasch commented on data-index/data-index-service/src/main/java/org/kie/kogito/index/vertx/VertxRouterSetup.java:
```diff
@@ -39,18 +47,44 @@
     Boolean authEnabled;
 
+    @Inject
+    @ConfigProperty(name = "kogito.dataindex.vertx-graphql.ui.path", defaultValue = "/graphiql")
```
done

nmirasch commented on data-index/data-index-service/src/test/java/org/kie/kogito/index/auth/MultiTenantResolverTest.java:
```diff
@@ -0,0 +1,70 @@
+    @Test
+    void resolveGraphiqlTenantTest() {
+        lenient().when(requestMock.path()).thenReturn(GRAPH_UI_PATH);
+        assertThat(GRAPH_UI_TENANT.equals(multiTenantResolver.resolve(routingContextMock)));
```
updated

nmirasch commented on data-index/data-index-service/src/test/java/org/kie/kogito/index/auth/MultiTenantResolverTest.java:
```diff
@@ -0,0 +1,70 @@
+    @InjectMocks
+    MultiTenantResolver multiTenantResolver;
+
+    @Before
```
updated

nmirasch commented on data-index/data-index-service/src/test/java/org/kie/kogito/index/auth/MultiTenantResolverTest.java:
```diff
@@ -0,0 +1,70 @@
+    }
+
+    @Test
+    void resolveGraphiqlTenantTest() {
```
added

nmirasch commented:
@Sgitario  I added unit test for VertxRouterSetup and added some of your suggestions. Could you take another look? Thanks!

Sgitario commented on data-index/data-index-service/src/test/java/org/kie/kogito/index/auth/MultiTenantResolverTest.java:
```diff
@@ -0,0 +1,76 @@
+        givenTenantIsConfigured();
+    }
+
+    private void givenTenantIsConfigured() {
```
I think private methods should go at the bottom after public and protected methods.

nmirasch replied:
@Sgitario Moved, Thanks!

nmirasch commented:
@cristianonicolai @jstastny-cz  could you take a look? thanks!

nmirasch commented:
jenkins please retest this

nmirasch commented:
jenkins please retest this

cristianonicolai commented:
jenkins please retest this

cristianonicolai commented on data-index/data-index-service/src/main/java/org/kie/kogito/index/vertx/VertxRouterSetup.java:
```diff
@@ -39,18 +47,47 @@
+            multiMap.add(HttpHeaders.ACCESS_CONTROL_ALLOW_ORIGIN, "*");
+            multiMap.add(HttpHeaders.ACCESS_CONTROL_ALLOW_CREDENTIALS, "true");
+            multiMap.add(HttpHeaders.ACCESS_CONTROL_ALLOW_HEADERS, "Authorization, Content-Type, Accept");
+            multiMap.add("Bearer-token", "Bearer " + token);
```
Do we also need this one, isnt just the `AUTHORIZATION` header the one used to authenticate?

nmirasch replied:
I have simplified that headers, just passing the Authorization with the current logged user token.

cristianonicolai commented on data-index/data-index-service/src/main/java/org/kie/kogito/index/vertx/VertxRouterSetup.java:
```diff
@@ -39,18 +47,47 @@
+            MultiMap multiMap = MultiMap.caseInsensitiveMultiMap();
+            String token = getCurrentAccessToken(rc);
+            if (token.isEmpty()) {
+                token = rc.request().getHeader(HttpHeaders.AUTHORIZATION);
```
@nmirasch when you get token like this wouldnt it contain the starting string "Bearer " ?

cristianonicolai replied:
do we actually need or is there a situation where the user could get to the GraphiQL UI and not be logged in or have access to it via a token in the authorization header? Perhaps we can simply send the current access token all the time?

nmirasch replied:
agree, simplified :)

cristianonicolai commented on data-index/data-index-service/src/test/java/org/kie/kogito/index/auth/MultiTenantResolverTest.java:
```diff
@@ -0,0 +1,77 @@
+
+    @BeforeEach
+    public void setup(){
+        lenient().when(routingContextMock.request()).thenReturn(requestMock);
```
@nmirasch just curious why do we need to use lenient in these tests?

cristianonicolai commented on data-index/data-index-service/src/test/java/org/kie/kogito/index/service/KeycloakIntegrationIndexingServiceIT.java:
```diff
@@ -31,18 +31,18 @@
         //alice only have role User, resource is forbidden
-        given().auth().oauth2(getAccessToken("alice"))
-                .when().get("/graphiql/")
+        given().contentType(ContentType.JSON).body("{ \"query\" : \"{ProcessInstances{ id } }\" }")
```
@nmirasch I got confused here, are we just running the same get request twice?

nmirasch replied:
yep double check removed

cristianonicolai commented on data-index/data-index-service/src/test/java/org/kie/kogito/index/service/KeycloakIntegrationIndexingServiceIT.java:
```diff
@@ -65,27 +61,38 @@ public void testNoTokenProvided() {
+                .when().get("/graphiql/")
+                .then()
+                .statusCode(200);
+        given().when().get("/other/")
```
not sure I get the point to check this endpoint as it will always return 404, with or without security :)

cristianonicolai commented on data-index/data-index-service/src/test/java/org/kie/kogito/index/vertx/VertxRouterSetupTest.java:
```diff
@@ -0,0 +1,116 @@
+    @Test
+    public void testAuthEnabledTrue() {
+        vertxRouterSetup.authEnabled = true;
+        vertxRouterSetup.graphUIPath="/graphiql";
```
```suggestion
        vertxRouterSetup.graphUIPath = "/graphiql";
```

cristianonicolai commented on data-index/data-index-service/src/test/java/org/kie/kogito/index/vertx/VertxRouterSetupTest.java:
```diff
@@ -0,0 +1,116 @@
+    @Test
+    public void testAuthEnabledFalse() {
+        vertxRouterSetup.authEnabled = false;
+        vertxRouterSetup.graphUIPath="/graphiql";
```
```suggestion
        vertxRouterSetup.graphUIPath = "/graphiql";
```

cristianonicolai commented on data-index/data-index-service/src/test/java/org/kie/kogito/index/vertx/VertxRouterSetupTest.java:
```diff
@@ -0,0 +1,116 @@
+        vertxRouterSetup.graphUIPath="/graphiql";
+        vertxRouterSetup.setupRouter(routerMock);
+
+        verify(vertxRouterSetup,never()).addGraphiqlRequestHeader(any());
```
```suggestion
        verify(vertxRouterSetup, never()).addGraphiqlRequestHeader(any());
```

nmirasch replied:
done

cristianonicolai commented on data-index/data-index-service/src/test/java/org/kie/kogito/index/vertx/VertxRouterSetupTest.java:
```diff
@@ -0,0 +1,116 @@
+        verify(graphiQLHandlerMock).graphiQLRequestHeaders(functionCaptor.capture());
+        functionCaptor.getValue().apply(routingContextMock);
+
+        ArgumentCaptor<MultiMap> multimapCaptor =ArgumentCaptor.forClass(MultiMap.class);
```
```suggestion
        ArgumentCaptor<MultiMap> multimapCaptor = ArgumentCaptor.forClass(MultiMap.class);
```

cristianonicolai commented on data-index/data-index-service/src/test/java/org/kie/kogito/index/vertx/VertxRouterSetupTest.java:
```diff
@@ -0,0 +1,116 @@
+
+        vertxRouterSetup.addGraphiqlRequestHeader(graphiQLHandlerMock);
+
+        ArgumentCaptor<Function<RoutingContext, MultiMap>> functionCaptor =ArgumentCaptor.forClass(Function.class);
```
```suggestion
        ArgumentCaptor<Function<RoutingContext, MultiMap>> functionCaptor = ArgumentCaptor.forClass(Function.class);
```

cristianonicolai commented on data-index/data-index-service/src/test/java/org/kie/kogito/index/vertx/VertxRouterSetupTest.java:
```diff
@@ -0,0 +1,116 @@
+    }
+
+    @Test
+    public void testAddGraphiqlRequestHeader() {
```
@nmirasch I think it makes sense to also test when the token is provided by the request headers. perhaps a new test method?

cristianonicolai commented on data-index/data-index-service/src/test/java/org/kie/kogito/index/vertx/VertxRouterSetupTest.java:
```diff
@@ -0,0 +1,116 @@
+        ArgumentCaptor<MultiMap> multimapCaptor =ArgumentCaptor.forClass(MultiMap.class);
+        verify(multiMapMock).addAll(multimapCaptor.capture());
+
+        assertThat(multimapCaptor.getValue().get(HttpHeaders.AUTHORIZATION).equals("Bearer " + token));
```
```suggestion
        assertThat(multimapCaptor.getValue().get(HttpHeaders.AUTHORIZATION)).isEqualTo("Bearer " + token);
```

nmirasch commented on data-index/data-index-service/src/main/java/org/kie/kogito/index/vertx/VertxRouterSetup.java:
```diff
@@ -39,18 +47,47 @@
+            multiMap.add(HttpHeaders.ACCESS_CONTROL_ALLOW_CREDENTIALS, "true");
+            multiMap.add(HttpHeaders.ACCESS_CONTROL_ALLOW_HEADERS, "Authorization, Content-Type, Accept");
+            multiMap.add("Bearer-token", "Bearer " + token);
+            rc.request().headers().addAll(multiMap);
```
yep, good catch absolutely right, Thanks!

nmirasch commented on data-index/data-index-service/src/main/java/org/kie/kogito/index/vertx/VertxRouterSetup.java:
```diff
@@ -39,18 +47,47 @@
+        if (Boolean.TRUE.equals(authEnabled)) {
+            addGraphiqlRequestHeader(graphiQLHandler);
         }
+        router.route(graphUIPath +"*").handler(graphiQLHandler);
```
done

nmirasch commented:
@cristianonicolai Thanks for the review. I have simplified the graphiql request header just to include the authorization one, and removed the code of retrieving the token from request headers regarding the entry point is the web app that is in charge to set that quarkus user with the credentials. And, finally I passed the reformat on files. Please could you take another look here?

Sgitario commented:
@nmirasch can you add the next into the [kogito-realm.json](https://github.com/kiegroup/kogito-apps/blob/82a4331f72edc512219a2013656bf821ce9fa6dd/data-index/data-index-service/src/test/resources/kogito-realm.json#L513)?

```
"redirectUris": [
        "*"
      ],
```

nmirasch commented:
@Sgitario  @cristianonicolai Renamed the keycloak client to kogito-service with the aim to reuse that to other kogito services and not to have to define a client to every one. I have updated the kogito-realm and added that to kogito-apps/config as a generic auth config  resource. Added some logic to allow "/graphiql" and "/graphiql/" as entry points to graphql UI. 
Could you  take a look and let me know what you think? Thanks!!

cristianonicolai commented on data-index/data-index-service/src/main/java/org/kie/kogito/index/vertx/VertxRouterSetup.java:
```diff
@@ -39,18 +47,46 @@
+                .orElse(StringUtils.EMPTY);
+    }
+
+    private String getGraphUIAllowedPath() {
```
@nmirasch Can this be a reusable utils function instead of duplicating it?

nmirasch replied:
@cristianonicolai  yes, I thought that but I'm not sure where makes sense to have it. I had the temptation to add a property for that.

cristianonicolai replied:
to be honest, I dont see much value to allow customizing this endpoint.

nmirasch replied:
@cristianonicolai  agree, I keep that as it was before

cristianonicolai commented on data-index/data-index-service/src/main/java/org/kie/kogito/index/auth/MultiTenantResolver.java:
```diff
@@ -0,0 +1,53 @@
+public class MultiTenantResolver implements TenantResolver {
+
+    @Inject
+    @ConfigProperty(name = "kogito.data-index.vertx-graphql.ui.path", defaultValue = "/graphiql/")
```
should the default be just _/graphiql_ ?

nmirasch replied:
updated

Sgitario commented on data-index/data-index-service/src/main/java/org/kie/kogito/index/vertx/VertxRouterSetup.java:
```diff
@@ -39,18 +47,46 @@
+        router.route().handler(StaticHandler.create());
+        router.route().handler(FaviconHandler.create());
+        router.route(getGraphUIAllowedPath()).handler(ctx -> ctx.response().putHeader("location", graphUIPath).setStatusCode(302).end());
+        router.route("/").handler(ctx -> ctx.response().putHeader("location", graphUIPath).setStatusCode(302).end());
```
In order to avoid duplicate code here, does it make sense to have something like:

```
router.route(graphUIPath + "*").handler(setRootLocation());
router.route("/").handler(setRootLocation());
// ...

protected Handler<RoutingContext> setRootLocation() {
     return ctx -> ctx.response().putHeader("location", graphUIPath).setStatusCode(302).end();
}
```

nmirasch replied:
@Sgitario Regarding I just restored the previous served routes, This is not needed now, but thanks!

nmirasch commented on data-index/data-index-service/src/main/java/org/kie/kogito/index/auth/MultiTenantResolver.java:
```diff
@@ -0,0 +1,53 @@
+    @Override
+    public String resolve(RoutingContext context) {
+        if (graphUIPath != null
+                && graphUITenantId != null
```
@Sgitario yes, removed!

nmirasch commented:
@Sgitario @cristianonicolai updated! when you have the chance, could you give thiis a thy? thanks!!!

cristianonicolai commented on data-index/data-index-service/src/test/java/org/kie/kogito/index/auth/MultiTenantResolverTest.java:
```diff
@@ -0,0 +1,87 @@
+@ExtendWith(MockitoExtension.class)
+class MultiTenantResolverTest {
+
+    private static final String GRAPH_UI_PATH = System.getProperty("kogito.data-index.vertx-graphql.ui.path", "/graphiql/");
```
```suggestion
    private static final String GRAPH_UI_PATH = System.getProperty("kogito.data-index.vertx-graphql.ui.path", "/graphiql");
```

cristianonicolai commented on data-index/data-index-service/src/test/java/org/kie/kogito/index/service/KeycloakIntegrationIndexingServiceIT.java:
```diff
@@ -65,27 +56,43 @@ public void testNoTokenProvided() {
+        // Returning keycloak login page
+        given().when().get("/graphiql/")
+                .then()
+                .statusCode(200);
```
@nmirasch I think just this one needs some improvement. A 200 could also be returned if the auth is disabled right? Anything that checks that it actually reached the login page.

nmirasch commented:
@cristianonicolai  added assertion to check the login page returned
